# Gene family evolution in Todiramphus

```{r}
library(ape)
library(stringr)
library(phytools)
library(readxl)
library(stringr)
library(UpSetR)
library(stringr)
library(pbmcapply)
```

## Estimate error to account for incomplete assemblies (e.g., for B10k genomes)

```{r, eval=FALSE}
setwd("~/Downloads/CAFE")
system("python cafe/caferror.py -i kingfishers/cafe_error1.sh -d kingfishers/error -f 1 -v 0 -s 1")
# system("python cafe/caferror.py")

# Now run accounting for error
system("cafe cafe_king1.sh")
```

## Analysis

```{r}
setwd("~/Dropbox/Projects/King_genome")

# CAFE3- "An error value of Îµ = 0.1 means that in 90% of gene families, the observed size is equal to the true size, whereas in 10% of gene families, the observed size is either too large or too small (fig. 1A, C, and E)."

# Plot
tr <- read.tree("~/Dropbox/Projects/Coraciiform morphometrics/output/revbayes/data/coraci192-beast-fixed.tre")
tr <- drop.tip(tr, which(!(tr$tip %in% c('Todiramphus_chloris','Todus_mexicanus','Halcyon_senegalensis','Chloroceryle_aenea','Ceyx_cyanopectus'))))
source("/Users/chadeliason/R/sppabbrev.R")
tr$tip.label <- sppAbbrev(tr$tip)
tr$root.edge <- 25
maxage <- max(branching.times(tr))
tr <- bind.tip(tr, tip.label='taeGut', position=62-maxage, edge.length=62)
maxage <- max(branching.times(tr))
tr <- bind.tip(tr, tip.label='calAnn', position=67.4-maxage, edge.length=67.4)
plot(tr)
is.ultrametric(tr)
# write.tree(tr, file="~/Dropbox/Projects/King_genome/analysis-cafe/cafe-tree.phy")

# setwd("~/Dropbox/Projects/King_genome/analysis-cafe")

# system("cafe cafe_run1.sh")

# tr <- read.tree(file="cafe-tree.phy")
# plot(tr)


# raw <- readLines("report.txt.cafe")[1:10]
# raw

# raw <- readLines("report.txt.cafe")[1:10]

#-------------------------------------------------------------------------------
# EDIT FOR REVISIONS ON PAPER
# raw <- readLines("/Users/chadeliason/Downloads/CAFE/kingfishers/cafe_final_report.cafe")

# cafe5 -i Orthogroups.GeneCount_forCAFE5.txt -t coraci_tree.phy

resfile <- "/Users/chadeliason/Downloads/CAFE/kingfishers/error/cafe_final_report.cafe"
# resfile <- "/Users/chadeliason/Downloads/CAFE/kingfishers/cafe_final_report.cafe"

raw <- readLines(resfile)

# avg expansion
# avgexp <- as.numeric(str_match_all(raw[6], "[0-9\\.\\-]+")[[1]][, 1])

# caford <- rank(as.numeric(str_match_all(raw[4], "[0-9]+")[[1]][,1]))
caford <- as.character(as.numeric(str_match_all(raw[4], "[0-9]+")[[1]][,1]))
# names(avgexp) <- caford

tr <- read.tree(text=paste0(gsub("Tree:", "", raw[1]), ";"))

phyord <- c('0'=1,'2'=2,'4'=3,'6'=4,'8'= 5,'7'=13,'5'=12,'3'=11,'1'=10,'10'=6,'9'=9,'12'=7,'11'=8)

# Significant changes only
out <- read.delim(resfile, head=T, skip=9, sep="\t")
out$padj <- p.adjust(out[, 3], method="fdr")
# table(out$padj < 0.05)
keep <- out[, 3] < 0.05  # significant ones
table(keep)  # N = 1443 orthogroups

# sig <- as.data.frame(do.call(rbind, strsplit(gsub("\\(|\\)", "", out[keep, 4]), split=",")))
# plot(sig[, 1])
# abline(h=0.05, lty=2)

# Get family-wide P values of gene families expanded/contracted
res <- t(sapply(1:nrow(out), function(x) str_match_all(out[x, 2], "_(\\d+)\\:")[[1]][, 2] ))
res <- apply(res, 2, as.numeric)
colnames(res) <- str_match_all(raw[3], "<(\\d+)>")[[1]][, 2]
colnames(res) <- phyord[colnames(res)]  # 
res <- res[, order(as.numeric(colnames(res)))]
rownames(res) <- paste0("og", 1:nrow(res))

# Ids for tips and nodes to use later
tid <- as.character(1:Ntip(tr))
nid <- as.character((Ntip(tr)+1) : (Ntip(tr) + Nnode(tr)))

# Node-wise P values-
nodesig <- as.data.frame(do.call(rbind, strsplit(gsub("\\(|\\)", "", out[keep, 4]), split=",")))
colnames(nodesig) <- phyord[caford]
head(nodesig)

# Only keep certain OGs
# res.signif <- res  # just did this to check that results match up with CAFE output, they did
res.signif <- res[keep, ]

# Calculate changes along branches of a phylogeny
changes <- t(sapply(1:nrow(res.signif), function(x) {
	# x=1
	states <- res.signif[x, ]
	changes <- states[tr$edge[, 2]] - states[tr$edge[, 1]]
	pvals <- nodesig[x, match(names(changes), colnames(nodesig))]  # Viterbi P values from cafe
	ifelse(pvals < 0.05, changes, 0)
}))
rownames(changes) <- rownames(res.signif)
colnames(changes) <- tr$edge[, 2]

sum_nochange <- apply(changes, 2, function(x) sum(x == 0))  # no change
sum_expand <- apply(changes, 2, function(x) sum(x > 0))  # expand
sum_contract <- apply(changes, 2, function(x) sum(x < 0))  # contract

tot_expand <- apply(changes, 2, function(x) sum(x[x > 0]))  # expand
tot_contract <- apply(changes, 2, function(x) sum(x[x < 0]))  # contract

change_per_gene <- apply(changes, 2, sum) / nrow(res.signif)

pdf("figs/gene_expansion_err_signif_viter.pdf", width=3, height=6)
plot(tr, edge.col = ifelse(change_per_gene > 0, "red", "blue"), no.margin = TRUE, cex = 0.75)
edgelabels(text = round(change_per_gene, 3), frame = "none", adj = c(0.5, -0.5), cex = 0.7)
edgelabels(text = paste0("+", sum_expand, "/ -", sum_contract), frame = "none", adj = c(0.5, 1.5), cex = 0.7)
dev.off()

# Significantly expanded in todChl
picks <- names(which((changes[, "5"] > 0)))
picks <- as.numeric(str_extract(picks, "\\d+"))
ogs[picks,]
transcripts_expanded <- unlist(strsplit(ogs[picks, 8], ", "))  # Convert to list of transcript names
picks <- names(which((changes[, "5"] < 0)))
picks <- as.numeric(str_extract(picks, "\\d+"))
transcripts_contracted <- unlist(strsplit(ogs[picks, 8], ", "))

# Load positive selection/gene name annotations
gene_lookup <- read.csv("analysis-cafe/pos_selection_results_repeatmasked_20210504.csv", row=1)
gene_lookup$file <- gsub(":", "_", gene_lookup$file)

# Load orthogroup output
ogs <- read.delim("analysis-cafe/Orthogroups.tsv")

# Get gene name list
genes <- gene_lookup$finalgene[match(transcripts_expanded, gene_lookup$file)]
# setdiff(transcripts_expanded, gene_lookup$file)
# grep("0304580", gene_lookup$file,value=T)
# head(gene_lookup)
genes <- genes[!is.na(genes)]
genes <- sort(unique(genes))
clipr::write_clip(genes)  # put this into STRING network website for analysis

# Load interproscan annotation
# iprdat <- read.delim("analysis-cafe/proteins_gemoma_round1_interpro.tsv", head = FALSE)
iprdat <- read.delim("output/proteins_all_gemoma_round2.faa.tsv", head=FALSE)
iprdat$V1 <- gsub(":", "_", iprdat$V1)  # fix "aqu..." gene names
# setdiff(transcripts_expanded, iprdat$V1)
# intersect(transcripts_expanded, iprdat$V1)
# aquChr_transcript_ENSACCT00020012595_R1
# sort(unique(grep("aqu", iprdat$V1, value=T)))

# todiramphus green color (RGB) in map- 72	170	129	

# Get list of GO terms

# ss <- unique(iprdat[iprdat$V1 %in% transcripts_expanded & grepl("rhodopsin", iprdat$V13), "V13"])
# gene_lookup[gene_lookup$file %in% ss, ]
# grep("rhodopsin", iprdat$V13)

# grep("^COL", gene_lookup$finalgene, value=T)
# gene_lookup[grep("COL4A4", gene_lookup$finalgene), ]

iprdat[grep("galGal_rna-XM_015276986.2_R0", iprdat$V1), ]

grep("Coll", iprdat[match(transcripts_expanded, iprdat$V1), "V13"], value=T)

goterms1 <- iprdat$V14[match(transcripts_expanded, iprdat$V1)]
goterms1 <- goterms1[!is.na(goterms1) & goterms1!=""]
goterms1 <- strsplit(goterms1, "\\|")
# tmp <- table(unlist(goterms1))
# clipr::write_clip(data.frame(names(tmp), value=as.numeric(tmp)))
goterms1 <- sort(unique(unlist(goterms1)))
clipr::write_clip(unlist(goterms1)) # for use in REVIGO (Fig. 3B panel)
clipr::write_clip(unlist(goterms1), breaks=",") # for use in REVIGO (Fig. 3B panel)

# Genes associated with each term-

picks <- iprdat$V1 %in% transcripts_expanded & grepl("GO:0007186", iprdat$V14) # GPCR
iprdat[picks, ]

picks <- iprdat$V1 %in% transcripts_expanded & grepl("GO:0007186", iprdat$V14) & grepl("[Oo]psin", iprdat$V13)# GPCR
iprdat[picks, ]
sort(unique(iprdat$V1[picks]))  # 34 transcripts

# aquChr_transcript_ENSACCT00020003396 = OPN1SW (UV vision.. but when I blast the sequence I get red-sensitive), FASTKD2 (energy balance)
# aquChr_transcript_ENSACCT00020014047 = CYB561D2
# aquChr_transcript_ENSACCT00020015072 = N/A
# aquChr_transcript_ENSACCT00020015981 = P2RY6
# aquChr_transcript_ENSACCT00020022606 = N/A
# XM_008501222.2 = GPRC
# XM_008502237.2 = mas-related G-protein coupled receptor member H
# XM_008502244.2 = mas-related G-protein coupled receptor member H
# XM_030468285.1 = olfactory receptor 5AN1-like
# XM_030468338.1 = olfactory receptor 14I1-like
# calAnn_rna-XM_030468472.1 = olfactory receptor 10AG1-like
# XM_030468600.1 = uncharacterized protein
# XM_015272533.2 = 
# galGal XM_025145334.1 = olfactory receptor 14J1
# galGal XM_025145352.1 = olfactory receptor 14I1-like
# None for melUnd


gene_lookup$finalgene[match(unique(iprdat$V1[picks]), gene_lookup$file)]

aquChr_transcript_ENSACCT00020003396_R4

goterms2 <- iprdat$V14[match(transcripts_contracted, iprdat$V1)]
goterms2 <- goterms2[!is.na(goterms2) & goterms2!=""]
goterms2 <- strsplit(goterms2, "\\|")
goterms2 <- sort(unique(unlist(goterms2)))
clipr::write_clip(unlist(goterms2)) # for use in REVIGO (Fig. 3B panel)

# interproscan results 
head(iprdat)
iprnums <- iprdat[match(transcripts, iprdat$V1), "V12"]
iprnums <- iprnums[!is.na(iprnums) & iprnums!=""]
# head(subset(iprdat, V1 %in% transcripts & V4=="PANTHER"))
# clipr::write_clip(subset(iprdat, V1 %in% transcripts & V4=="PANTHER")[c('V1','V5')])
# clipr::write_clip(subset(iprdat, V1 %in% transcripts & V4=="PANTHER")$V5, breaks=",")

table(is.na(gene_lookup$finalgene[match(transcripts, gene_lookup$file)]))
table(is.na(iprdat$V14[match(transcripts, iprdat$V1)]))
table(is.na(iprdat$V5[match(transcripts, iprdat$V1)]))

pthr <- iprdat[match(transcripts, iprdat$V1), "V5"]

res1 <- dcGOR::dcEnrichment(iprnums, domain="InterPro", ontology="KW")
dcGOR::view(res1)



#-------------------------------------------------------------------------------

# OLD STUFF-

edgedat <- readxl::read_excel("analysis-cafe/cafe_edge_stats.xlsx", sheet=1)

# Compare species-specific error rates with assembly
plot(rate_err ~ rate_spperr, edgedat)
abline(a=0, b=1, lty=2)

plot(expand_spperr ~ expand_err, edgedat)
abline(a=0, b=1)

xx <- edgedat$rate_spperr

# minmax <- function(x) {
#   (x-min(x)) / (max(x-min(x)))
# }

head(edgedat)

pdf("figs/gene_expansion_err.pdf", width=3, height=6)
plot(tr, edge.col=ifelse(xx>0, "red", "blue"), no.margin=T, cex=0.75)
edgelabels(text=round(xx,3), frame="none", adj=c(0.5, -.5), cex=.7)
edgelabels(text=paste0("+", edgedat$expand_spperr, "/ -", edgedat$decrease_spperr), frame="none", adj=c(0.5, 1.5), cex=.7)
dev.off()

# edgelabels(text=edgedat$Expand)
# edgelabels(text=edgedat$nDecrease)
# colorRamp(colors=rainbow(25))(1)

# Load gene families
ogs <- read.delim("analysis-cafe/Orthogroups.tsv")

x <- strsplit(ogs[,8], ",")

# Find families that have expanded in todChl
res <- read.delim("analysis-cafe/report.txt.cafe", skip=10)
trees <- res[,2]

# Find todChl-halSen branch in tree
m <- str_match(trees, "halSen_(\\d+).*?todChl_(\\d+).*?\\)_(\\d+)")

# Get number of unique gene families expanded in todChl
x1 <- as.numeric(m[,2])  # halSen
x2 <- as.numeric(m[,3])  # todChl
x3 <- as.numeric(m[,4])  # ancestor

table(x2&!x1)

# Get number of unique gene families expanded in ceyCya
m2 <- str_match(trees, "ceyCya\\_(\\d+)\\:.*?(\\d+)\\:16\\.7776")
table(m2[,2]>m2[,3])

unlist(strsplit(ogs[m2[,2]>m2[,3], 8], ","))

intersect(which(m2[,2]>m2[,3]), which(x2>x3))
table(x2>x3 & m2[,2]>m2[,3])

ogstats <- read.delim("output/orthofinder_may06/Orthogroups.GeneCount.tsv")
names(ogstats)[2:8] <- c('todMex','halSen','chlAen','ceyCya','calAnn','taeGut','todChl')

idx <- which(x2>x3 & ogstats[,"ceyCya"]!=0)

# Find todChl genes in orthogroups gained in todChl relative to MRCA with halSen
transcripts <- unlist(strsplit(ogs[which(x2 > x3), 8], ", "))
# transcripts2 <- unlist(strsplit(ogs[which(x2 > x3 & x1 < x3), 8], ", "))
# transcripts3 <- unlist(strsplit(ogs[idx, 8], ", "))

setdat <- ogstats[,2:8]
setdat <- cbind(set='a',setdat)
setdat[,-1] <- ifelse(setdat[,-1] >= 1, 1, 0)

# Upset plot
pdf("figs/todChl_upset.pdf", width=6.5, height=4.5)
# par(mex=0.75,mgp=c(1.5,.5,0),ps=8)
# upset(subset(setdat, x2!=0 & x1==0), nsets=7, sets=c('todMex','ceyCya','chlAen','halSen','todChl','taeGut','calAnn'), nintersects=35, order.by="freq", mainbar.y.label="Number of orthogroups", keep.order=TRUE)
upset(setdat, nsets=7, sets=c('todMex','ceyCya','chlAen','halSen','todChl','taeGut','calAnn'), nintersects=35, order.by="freq", mainbar.y.label="Number of orthogroups", keep.order=TRUE,
		text.scale=c(1,1,1,1,1,.75))
# upset(subset(setdat, todChl==1), nsets=7, sets=c('todMex','ceyCya','chlAen','halSen','todChl','taeGut','calAnn'), nintersects=35, order.by="freq", mainbar.y.label="Number of orthogroups", keep.order=TRUE)
# upset(setdat, nsets=7, nintersects=NA, order.by="freq", mainbar.y.label="Number of orthogroups", queries=list(list(query=intersects, params=list(c("todChl","chlAen","halSen"), c("todChl","chlAen")))))
dev.off()


# Lookup genes from GFF files used in gemoma-
gene_lookup <- read.csv("analysis-cafe/pos_selection_results_repeatmasked_20210504.csv", row=1)
# gene_lookup <- read.csv("paml/pos_selection_results_repeatmasked_20210504.csv", row=1)

picks <- gene_lookup$file[grep("TAS2R7", gene_lookup$finalgene)]
plot(sapply(strsplit(as.character(ogs[grep(picks[1], ogs[,8]), ]), "\\,"), length))

picks <- gene_lookup$file[grep("TAS2R40", gene_lookup$finalgene)]
plot(sapply(strsplit(as.character(ogs[grep(picks[1], ogs[,8]), ]), "\\,"), length))

picks <- gene_lookup$file[grep("TAS2R114", gene_lookup$finalgene)]
plot(sapply(strsplit(as.character(ogs[grep(picks[1], ogs[,8]), ]), "\\,"), length))

picks <- gene_lookup$file[grep("TAS1R3", gene_lookup$finalgene)]
ogs[grep(picks[1], ogs[,8]), ]

picks <- gene_lookup$file[grep("MC1R", gene_lookup$finalgene)]
ogs[grep(picks[1], ogs[,8]), ]


# scp celiason@10.100.111.202:~/uce-alcedinidae/annotations/gemoma/round2/uniprot.all.mmseqs ~/Downloads/uniprot_todChl_annot_gemoma_round2.txt
annot <- read.delim("~/Downloads/uniprot_todChl_annot_gemoma_round2.txt", head=F)
annot$V1 <- gsub(":", "_", annot$V1)

setdiff(res, annot$V1)  # e.g., aquChr_transcript_ENSACCT00020000837_R0

# Get gene names
genes <- gene_lookup$finalgene[match(transcripts, gene_lookup$file)]
genes <- genes[!is.na(genes)]
genes <- sort(unique(genes))
clipr::write_clip(genes)  # put this into STRING network website for analysis

# GO terms
# TODO update with round2 functional annotations
tmp <- read.delim("analysis-cafe/proteins_gemoma_round1_interpro.tsv", head = FALSE)

goterms <- tmp$V14[match(res, tmp$V1)]
goterms <- goterms[!is.na(goterms) & goterms!=""]
goterms <- strsplit(goterms, "\\|")
goterms <- sort(unique(unlist(goterms)))
clipr::write_clip(unlist(goterms)) # for use in REVIGO (Fig. 3B panel)

# interproscan results 
iprnums <- tmp[match(transcripts, tmp$V1), "V12"]
iprnums <- iprnums[!is.na(iprnums) & iprnums!=""]

res1 <- dcEnrichment(iprnums, domain="InterPro", ontology="GOBP")
view(res1)

# Functional annotation of genes
setwd("~/uce-alcedinidae/ortho/proteomes/primary_transcripts/OrthoFinder/Results_May06/WorkingDirectory/OrthoFinder/Results_May06")
dat <- read.delim("Orthogroups/Orthogroups.GeneCount.tsv")

# Orthogroups specific to todChl-
picks <- as.character(dat[apply(dat[,2:7], 1, sum)==0, "Orthogroup"])
picks <- paste0("Orthogroup_Sequences/", picks, ".fa")
seqs <- lapply(picks, readLines)
# tmp <- read.csv("~/uce-alcedinidae/paml/pos_selection_results_repeatmasked_20210504.csv")
# tmp$finalgene[match(gsub(">", "", grep(">", unlist(seqs), value=T)), tmp$file)]
cat(unlist(seqs), sep="\n", file="~/uce-alcedinidae/ogs_todChl.faa")

# 1) Functional annotations of this set of 404 genes using MMSEQS-
# mmseqs easy-rbh --threads 12 ogs_todChl.faa annotations/gemoma/uniprot_sprot.fasta ogs_todChl_uniprot.mmseqs tmp
# head ogs_todChl_uniprot.mmseqs

# Gene list



# 2) Functional annotation using InterProScan-
alias ips=/home/FM/celiason/uce-alcedinidae/interproscan-5.47-82.0/interproscan.sh
# ipsJar=/home/FM/celiason/uce-alcedinidae/interproscan/interproscan-5.47-82.0/interproscan-5.jar
# Real run-
# proteins.fasta file was produced by MAKER as standard output
# ips -i ogs_todChl.faa -goterms -f tsv
# head ogs_todChl.faa_1.tsv

# Load IPR annotated genes
dat <- read.delim("output/ogs_todChl.faa_1.tsv", head=F)

# Run enrichment analysis
library(dcGOR)
res <- dcEnrichment(dat$V12, domain="InterPro", ontology="GOBP")

# Visualize as a graph
g <- visEnrichment(res, node.info="full_term_name")
g2 <- visEnrichment(res, node.info="full_term_name", num_top_nodes=10)

# Export terms for REVIGO analysis
tops <- view(res, 125)
cat(tops$term_id, sep="\n")

view(res)

dat[grep("1752", dat$V12), ]
dat[grep("5844", dat$V12), ]
dat[grep("5845", dat$V12), ]
dat[grep("16055", dat$V12), ]
dat[grep("6001", dat$V12), ]

# R code to plot in REVIGO-

# A treemap R script produced by the Revigo server at http://revigo.irb.hr/
# If you found Revigo useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# author: Anton Kratz <anton.kratz@gmail.com>, RIKEN Omics Science Center, Functional Genomics Technology Team, Japan
# created: Fri, Nov 02, 2012  7:25:52 PM
# last change: Fri, Nov 09, 2012  3:20:01 PM

# -----------------------------------------------------------------------------
# If you don't have the treemap package installed, uncomment the following line:
# install.packages( "treemap" );
library(treemap) 								# treemap package by Martijn Tennekes

# Set the working directory if necessary
# setwd("C:/Users/username/workingdir");

# --------------------------------------------------------------------------
# Here is your data from Revigo. Scroll down for plot configuration options.

revigo.names <- c("term_ID","description","frequency","uniqueness","dispensability","representative");
revigo.data <- rbind(c("GO:0008150","biological_process",100,1,0,"biological_process"),
c("GO:0008152","metabolic process",62.693911296527,1,0,"metabolic process"),
c("GO:0009628","response to abiotic stimulus",0.603634500970981,0.925217811316376,0,"response to abiotic stimulus"),
c("GO:0009416","response to light stimulus",0.188810302859614,0.843634660246174,0.30112543,"response to abiotic stimulus"),
c("GO:0051716","cellular response to stimulus",11.0866663899183,0.89014938374801,0.46891116,"response to abiotic stimulus"),
c("GO:0009987","cellular process",78.0376878191047,1,0,"cellular process"),
c("GO:0043933","protein-containing complex subunit organization",1.6681875395806,0.935840927896943,0,"protein-containing complex subunit organization"),
c("GO:0006996","organelle organization",4.10954278773561,0.929874097335854,0.66494786,"protein-containing complex subunit organization"),
c("GO:0050794","regulation of cellular process",19.8740720541611,0.97184669100809,0,"regulation of cellular process"),
c("GO:0050896","response to stimulus",13.7208133053304,1,0,"response to stimulus"),
c("GO:0051179","localization",18.7941114253392,1,0,"localization"),
c("GO:0065007","biological regulation",23.3881336449155,1,0,"biological regulation"),
c("GO:1901135","carbohydrate derivative metabolic process",7.03889068001952,0.911362892805519,0,"carbohydrate derivative metabolic process"),
c("GO:1901360","organic cyclic compound metabolic process",22.4511996998524,0.878150523573222,0.12633896,"carbohydrate derivative metabolic process"),
c("GO:0043170","macromolecule metabolic process",34.4366477094063,0.85874161483324,0.20469617,"carbohydrate derivative metabolic process"),
c("GO:1901576","organic substance biosynthetic process",23.3073421911027,0.837907478511037,0.20773315,"carbohydrate derivative metabolic process"),
c("GO:1901564","organonitrogen compound metabolic process",32.1987066241535,0.841282333392547,0.23825687,"carbohydrate derivative metabolic process"),
c("GO:0044267","cellular protein metabolic process",15.254359469727,0.823898453789706,0.2866446,"carbohydrate derivative metabolic process"),
c("GO:0006139","nucleobase-containing compound metabolic process",18.7768975137146,0.83816886145711,0.30786996,"carbohydrate derivative metabolic process"),
c("GO:0034641","cellular nitrogen compound metabolic process",26.2084518110319,0.83172830309132,0.34939378,"carbohydrate derivative metabolic process"),
c("GO:0019538","protein metabolic process",19.8225173210066,0.837878750066921,0.43922804,"carbohydrate derivative metabolic process"),
c("GO:0044260","cellular macromolecule metabolic process",25.5354976543922,0.808874814264069,0.46854268,"carbohydrate derivative metabolic process"),
c("GO:0009059","macromolecule biosynthetic process",11.1039548744453,0.840549362958524,0.58291612,"carbohydrate derivative metabolic process"),
c("GO:0016192","vesicle-mediated transport",1.41412180423006,0.868094089715752,0.01304146,"vesicle-mediated transport"),
c("GO:0051649","establishment of localization in cell",2.04442026072841,0.861975425874124,0.34310922,"vesicle-mediated transport"),
c("GO:0051641","cellular localization",2.42716153907533,0.861405482004253,0.35682686,"vesicle-mediated transport"),
c("GO:0006812","cation transport",3.53439099252141,0.850009956472001,0.38656461,"vesicle-mediated transport"),
c("GO:0006810","transport",17.9971715326728,0.827491774029602,0.5528133,"vesicle-mediated transport"),
c("GO:0006811","ion transport",5.1712786278122,0.856725097409823,0.59779702,"vesicle-mediated transport"),
c("GO:0071702","organic substance transport",5.81485106092797,0.85425991410083,0.61317491,"vesicle-mediated transport"),
c("GO:0071840","cellular component organization or biogenesis",9.05104158872763,0.981371738760136,0.0167685,"cellular component organization or biogenesis"),
c("GO:0044281","small molecule metabolic process",15.7742278866694,0.9081532110822,0.07740644,"small molecule metabolic process"),
c("GO:0009058","biosynthetic process",24.4296768002127,0.894418865175645,0.10698742,"small molecule metabolic process"),
c("GO:0006807","nitrogen compound metabolic process",45.3621081991498,0.866258630750214,0.15835912,"small molecule metabolic process"),
c("GO:0044238","primary metabolic process",49.9247269409607,0.860620178906147,0.23456847,"small molecule metabolic process"),
c("GO:0071704","organic substance metabolic process",55.5008618556042,0.853950970497847,0.27143659,"small molecule metabolic process"),
c("GO:0044237","cellular metabolic process",51.5163612533451,0.848707235179424,0.27824028,"small molecule metabolic process"),
c("GO:0006796","phosphate-containing compound metabolic process",13.8765960983292,0.85407087781059,0.091154,"phosphate-containing compound metabolic process"),
c("GO:0006793","phosphorus metabolic process",14.1931042105766,0.875515013458844,0.16376197,"phosphate-containing compound metabolic process"),
c("GO:0046483","heterocycle metabolic process",21.6400578122282,0.859605123703167,0.18465362,"phosphate-containing compound metabolic process"),
c("GO:0006725","cellular aromatic compound metabolic process",21.6625912575535,0.859560879108118,0.21016732,"phosphate-containing compound metabolic process"));

stuff <- data.frame(revigo.data);
names(stuff) <- revigo.names;

stuff$frequency <- as.numeric( as.character(stuff$frequency) );
stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
stuff$dispensability <- as.numeric( as.character(stuff$dispensability) );

# by default, outputs to a PDF file
pdf( file="revigo_treemap.pdf", width=16, height=9 ) # width and height are in inches

# check the tmPlot command documentation for all possible parameters - there are a lot more
treemap(
  stuff,
  index = c("representative","description"),
  vSize = "uniqueness",
  type = "categorical",
  vColor = "representative",
  title = "Revigo TreeMap",
  inflate.labels = FALSE,      # set this to TRUE for space-filling group labels - good for posters
  lowerbound.cex.labels = 0,   # try to draw as many labels as possible (still, some small squares may not get a label)
  bg.labels = "#CCCCCCAA",   # define background color of group labels
								 # "#CCCCCC00" is fully transparent, "#CCCCCCAA" is semi-transparent grey, NA is opaque
  position.legend = "none"
)

dev.off()


#

sort(unique(subset(dat, V12 %in% unique(unlist(strsplit(tops$members, ","))))$V13))

picks <- tops$members[1]
subset(dat, V12 %in% strsplit(picks, ",")[[1]])$V13 %>% unique

picks <- tops$members[2]
subset(dat, V12 %in% strsplit(picks, ",")[[1]])$V13 %>% unique

picks <- tops$members[tops$term_name=="single-organism transport"]
subset(dat, V12 %in% strsplit(picks, ",")[[1]])$V13

picks <- tops$members[tops$term_name=="phosphate-containing compound metabolic process"]
unique(subset(dat, V12 %in% strsplit(picks, ",")[[1]])$V13)

subset(dat, V12 %in% c("IPR005844","IPR005845","IPR016055","IPR006001"))
# Alpha-D-phosphohexomutase I, II, I/II/III
# Carbohydrate kinase, thermoresistant glucokinase

subset(dat, V12 %in% c("IPR001752","IPR027640","IPR026983","IPR026815"))
# Dynein heavy chain
# Kinesin motor domain, Kinesin-related

jpg("figs/network_ogs.jpg", width=7, height=7)
plot(g, font=16)
dev.off()
```


Code graveyard-

```{r, eval=FALSE}
# library(igraph)
# class(g)
# dcConverter(res, from="Eoutput", to="igraph")
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("enrichplot")
# library(enrichplot)
# emapplot(res)
# emapplot(edo, pie_scale=1.5)

# Junk function?
x <- as.character(nms$file[3])
i=4
i=6473
# i=10395
nms[i, ]
x <- as.character(nms$file[1])

getGeneName <- function(x) {
	if (substr(x, 1, 6) == "aquChr") {
		id <- str_match(x, "(ENSACCT\\d+)")[2]
		raw <- system(paste0("grep ", id, " annotations/data/Aquila_chrysaetos_chrysaetos.bAquChr1.2.101.gff3"), intern=T)
		out <- str_match(raw, "mRNA.*?Name=([A-Z0-9]+)")[, 2]
		return(unique(na.omit(out)))
	} else if (substr(x, 1, 6) == "calAnn") {
		id <- str_match(x, "(rna-XM_\\d+)")[2]
		if (is.na(id)) {
			next
		}
		raw <- system(paste0("grep ", id, " annotations/data/GCF_003957555.1_bCalAnn1_v1.p_genomic.gff"), intern=T)
		out <- str_match(raw, "gene=(.*?);")[, 2]
		unique(na.omit(out))
	} else if (substr(x, 1, 6) == "galGal") {
		id <- str_match(x, "(rna-[NX]M_\\d+)")[2]
		if (is.na(id)) {
			next
		}
		raw <- system(paste0("grep ", id, " annotations/data/GCF_000002315.6_GRCg6a_genomic.gff"), intern=T)
		out <- str_match(raw, "gene=(.*?);")[, 2]
		unique(na.omit(out))
	} else {
		return(NA)
	}
}

# args(pbmclapply)

nms_new <- pbmclapply(nms$file, getGeneName, mc.cores=24, mc.preschedule=FALSE)

idx <- unlist(lapply(nms_new, function(x) length(x)==0 || is.na(x)))

nms_new[idx] <- "unknown"
nms$genematch <- unlist(nms_new)
head(nms)
nms[1:5000, c('file','finalgene','genematch')]
nms$genematch[grep("Error", nms$genematch)] <- "unknown"
nms$genematch[is.na(nms$finalgene)]

write.csv(nms, file="paml/pos_selection_results_repeatmasked_20210504_matched.csv")
```

Copy interproscan annotations from phoebe server-

```sh
scp celiason@10.100.111.202:~/uce-alcedinidae/annotations/gemoma/round2/proteins_all.faa.tsv ~/Downloads/proteins_all_gemoma_round2.faa.tsv
```

Now analyze these results-

```{r}
lookup <- read.delim("~/Downloads/proteins_all_gemoma_round2.faa.tsv", head=F, sep="\t")

tmp <- readLines("~/Downloads/CAFE/kingfishers2/run1_results.cafe")
ogs <- read.delim("~/Downloads/CAFE/kingfishers2/Orthogroups.GeneCount_edited.tsv")
pvals <- sapply(tmp$V4, function(x) as.numeric(str_match_all(x, "[e0-9\\.\\-]+")[[1]][, 1]))
idx <- setNames(which(pvals[, 2] < 0.05), NULL)
res <- unlist(strsplit(ogs[idx, 8], ", "))

# lookup[match(res, lookup$V1), ]
# grep("Taste", lookup$V6, value=T)
goterms <- lookup$V14[match(res, lookup$V1)]
goterms <- goterms[!is.na(goterms) & goterms!=""]
goterms <- strsplit(goterms, "\\|")
goterms <- sort(unique(unlist(goterms)))
clipr::write_clip(unlist(goterms)) # 42 unique GO terms
```
